### PR TITLE
Adds Psychologist Features

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Adds Psychologist interactions to the SM:
Being near the sm when it's on will cause it to calm the fuck down.
It will:
Take less damage, and heal twice as fast. Because a good talk helps you vent.
Output much less heat and gas, and lower its power output. As a shoulder to cry on can calm you down.
Makes the Psychologist immune to it's hallucination effects. Because mental health workers can't allow themselves to be effected by their patients anger, they can only help

## Why It's Good For The Game

Feature man good.
(I spent 4 days on a meme ok, filters are insane man)
Here's a high quality non nightcode video:
[perfection](https://cdn.discordapp.com/attachments/326831214667235328/697343252742930462/2020-04-07_23-35-26.mp4)

## Changelog
:cl:
add: Added Psychologist features to the SM
/:cl:

### CODE REVIEW

**Comment by wesoda25** [comment]
Finally some good fucking content

**Comment by Ryll-Ryll** [comment]
Psychologist passively having an aura of calmness around them like the chaplain and magic would be kinda funny

**Comment by Shadowflame909** [comment]
I say this should be unironically merged.

Not very different from what the other RP jobs have. Like the mime having invisible magic, and the curator being able to speak all languages, even seemingly impossible ones like swarmer or borg.

**Comment by bluetshirtguy** [comment]
Giving Psych a non RP that can be utilized here and there(or even as a part of a crazy sm setup) is an excellent idea, goood pr.

**Comment by XDTM** [comment]
The only thing i'd object to is tying it to roundstart assigned role

**Comment by trollbreeder** [comment]
april fools was 7 years ago...

**Comment by ExcessiveUseOfCobblestone** [comment]
lol

**Comment by Fikou** [comment]
the iconstate for shart is bugged

**Comment by MrDoomBringer** [comment]
while this is a meme, I think this is unironically one of the best examples of mutually beneficial job interactions we've seen, which is supposed to be a backbone of this game. I'm for a merge

**Comment by ExcessiveUseOfCobblestone** [comment]
It needs to be off the intrinsic job though since we support job switching (A Psych really couldn't be an engineer after this considering power output is negatively affected by their presence).

I'm unironically ok with merging this, weird interactions like this is cool (like the bee costume saving you from bees).

**Comment by LemonInTheDark** [comment]
> 
> 
> the iconstate for shart is bugged
Thank ya fikou

**Comment by LemonInTheDark** [comment]
> 
> 
> It needs to be off the intrinsic job though since we support job switching (A Psych really couldn't be an engineer after this considering power output is negatively affected by their presence).
> 
> I'm unironically ok with merging this, weird interactions like this is cool (like the bee costume saving you from bees).

I uhh I'm glad you like it. I've changed the job check to use mob.job, I'm not sure if that's the right way to go about this, let me know if I should do something else.

**Comment by Dorsisdwarf** [comment]
I think generally you tie it to a standard bit of role equipment that they spawn with/have in their locker

**Comment by optimumtact** [comment]
The coefficients are too high, these should be absolutely minor buffs, for flavour and fun

on the order of 10% maximum rate increases

**Comment by LemonInTheDark** [comment]
Can do, I'll shoot for something close to that, I want it to be a noticeable change.

File: code/modules/power/supermatter/supermatter.dm [hunk 1]
```diff
@@ -535,19 +546,22 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
...
 	for(var/mob/living/carbon/human/l in view(src, HALLUCINATION_RANGE(power))) // If they can see it without mesons on.  Bad on them.
-		if(!istype(l.glasses, /obj/item/clothing/glasses/meson))
+		if(l.job == "Psychologist")
+			toAdd = 0.05
+		else if(!istype(l.glasses, /obj/item/clothing/glasses/meson))
 			var/D = sqrt(1 / max(1, get_dist(l, src)))
 			l.hallucination += power * config_hallucination_power * D
 			l.hallucination = clamp(l.hallucination, 0, 200)
-
+	psyCoeff = max(min(toAdd + psyCoeff, 1),0)
```
Comment (LudwigVonChesterfield) [inline code/modules/power/supermatter/supermatter.dm:557]: couldn't you use CLAMP()?

Comment (LudwigVonChesterfield) [inline code/modules/power/supermatter/supermatter.dm:557]:
```suggestion
	psyCoeff = clamp(toAdd + psyCoeff, 0, 1)
```

File: code/modules/power/supermatter/supermatter.dm [hunk 2]
```diff
@@ -535,19 +546,22 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
...
 	for(var/mob/living/l in range(src, round((power / 100) ** 0.25)))
 		var/rads = (power / 10) * sqrt( 1 / max(get_dist(l, src),1) )
 		l.rad_act(rads)
 
 	//Transitions between one function and another, one we use for the fast inital startup, the other is used to prevent errors with fusion temperatures.
 	//Use of the second function improves the power gain imparted by using co2
-	power = max(power - min(((power/500)**3) * powerloss_inhibitor, power * 0.83 * powerloss_inhibitor),0)
+	power = max(power - min(((power/500)**3) * powerloss_inhibitor, power * 0.83 * powerloss_inhibitor) * (psyCoeff > 0 ? 1 - psyCoeff + 0.5 : 1),0)
```
Comment (LudwigVonChesterfield) [inline code/modules/power/supermatter/supermatter.dm:564]:
```suggestion
	power = max(power - min(((power/500)**3) * powerloss_inhibitor, power * 0.83 * powerloss_inhibitor) * (psyCoeff > 0 ? 1.5 - psyCoeff : 1), 0)
```

But 1.5 would look like such a magical number.

File: code/modules/power/supermatter/supermatter.dm [hunk 3]
```diff
@@ -510,7 +521,7 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
 			src.fire_nuclear_particle()        // Start to emit radballs at a maximum of 30% chance per tick
 
 		//Power * 0.55
-		var/device_energy = power * REACTION_POWER_MODIFIER
+		var/device_energy = power * REACTION_POWER_MODIFIER * (psyCoeff == 1 ? 0.5 : 1)
```
Comment (LudwigVonChesterfield) [inline code/modules/power/supermatter/supermatter.dm:524]: It seems like everywhere psyCoeff affects at a rate of "0.5". Why not make psyCoeff maximally go to 0.5(and also make a define like MAX_PSYCOEFF), but to mitigate the speed of getting to max change the change rate from 0.05 to 0.025?

File: code/modules/power/supermatter/supermatter.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/modules/power/supermatter/supermatter.dm:557]: yeet

File: code/modules/power/supermatter/supermatter.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/modules/power/supermatter/supermatter.dm:564]: Frankly I need to rewrite the math here, but I'll keep that in mind

File: code/modules/power/supermatter/supermatter.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/modules/power/supermatter/supermatter.dm:524]: I'd prefer to keep it between 0 and 1 for understandability.

**Comment by LemonInTheDark** [comment]
Fucking hell github

**Pull request closed without merging**